New ipq4019 device support: Google WiFi

computersforpeace opened the pull request:
This series adds support for Google WiFi. It's mostly a standard ipq4019 system, but it runs a Coreboot/Depthcharge bootloader and so utilizes the Chromium verified boot payload-packaging format.

See the commits for plenty of details. There's also a lot of info here: https://openwrt.org/inbox/toh/google/google_wifi

There are some non-upstream kernel patches in here to get SMP and reboot working. They're not absolutely required (so, the first part of the series could be merged without them), and I'm also open to tweaking them (e.g., adding a `compatible` property?) if desired. I may work on getting them upstream eventually; I've had some discussion with some of the maintainers.

computersforpeace committed:
ipq40xx: Add subtarget for Google WiFi (Gale)

Google WiFi (codename: Gale) is an IPQ4019-based AP, with 2 Ethernet
ports, 2x2 2.4+5GHz WiFi, 512 MB RAM, 4 GB eMMC, and a USB type C port.
In its stock configuration, it runs a Chromium OS-based system, but you
wouldn't know it, since you can only manage it via a "cloud" +
mobile-app system.

The "v2" label is coded into the bootloader, which prefers the
"google,gale-v2" compatible string. I believe "v1" must have been
pre-release hardware.

Note: this is *not* the Google Nest WiFi, released in 2019.

I include "factory.bin" support, where we generate a GPT-based disk
image with 2 partitions -- a kernel partition (using the custom "Chrome
OS kernel" GUID type) and a root filesystem partition. See below for
flashing instructions.

Sysupgrade is supported via recent emmc_do_upgrade() helper.

This is a subtarget because it enables different features
(FEATURES=boot-part rootfs-part) whose configurations don't make sense
in the "generic" target, and because it builds in a few USB drivers,
which are necessary for installation (installation is performed by
booting from USB storage, and so these drivers cannot be built as
modules, since we need to load modules from USB storage).

Flashing instructions
=====================

Documented here:
https://openwrt.org/inbox/toh/google/google_wifi

Note this requires booting from USB storage.

Features
========

I've tested:

 * Ethernet, both WAN and LAN ports
 * eMMC
 * USB-C (hub, power-delivery, peripherals)
 * LED0 (R/G/B)
 * WiFi (limited testing)
 * SPI flash
 * Serial console: once in developer mode, console can be accessed via
   the USB-C port with SuzyQable, or other similar "Closed Case
   Debugging" tools:
     https://chromium.googlesource.com/chromiumos/third_party/hdctools/+/master/docs/ccd.md#suzyq-suzyqable
 * Sysupgrade

Not tested:

 * TPM

Known not working:

 * Reboot: this requires some additional TrustZone / SCM
   configuration to disable Qualcomm's SDI. I have a proposal upstream,
   and based on IRC chats, this might be acceptable with additional DT
   logic:
     [RFC PATCH] firmware: qcom_scm: disable SDI at boot
     https://lore.kernel.org/linux-arm-msm/20200721080054.2803881-1-computersforpeace@gmail.com/
 * SMP: enabling secondary CPUs doesn't currently work using the stock
   bootloader, as the qcom_scm driver assumes newer features than this
   TrustZone firmware has. I posted notes here:
     [RFC] qcom_scm: IPQ4019 firmware does not support atomic API?
     https://lore.kernel.org/linux-arm-msm/20200913201608.GA3162100@bDebian/
 * There's a single external button, and a few useful internal GPIO
   switches. I haven't hooked them up.

The first two are fixed with subsequent commits.

Additional notes
================

Much of the DTS is pulled from the Chrome OS kernel 3.18 branch, which
the manufacturer image uses.

Note: the manufacturer bootloader knows how to patch in calibration data
via the wifi{0,1} aliases in the DTB, so while these properties aren't
present in the DTS, they are available at runtime:

  # ls -l
/sys/firmware/devicetree/base/soc/wifi@a*/qcom,ath10k-pre-calibration-data
  -r--r--r--    1 root     root         12064 Jul 15 19:11 /sys/firmware/devicetree/base/soc/wifi@a000000/qcom,ath10k-pre-calibration-data
  -r--r--r--    1 root     root         12064 Jul 15 19:11 /sys/firmware/devicetree/base/soc/wifi@a800000/qcom,ath10k-pre-calibration-data

Ethernet MAC addresses are similarly patched in via the ethernet{0,1} aliases.

Signed-off-by: Brian Norris <computersforpeace@gmail.com>

computersforpeace committed:
ipq40xx: Support Chromium OS image-type creation

See firmware-utils.git commits [1], which implemented the cros-vbutil
verified-boot payload-packing tool, and extended ptgen for the CrOS
kernel partition type. With these, it's now possible to package kernel +
rootfs to make disk images that can boot a Chrome OS-based system (e.g.,
Chromebooks, or even a few AP models).

Regarding PARTUUID= changes: Chromium bootloaders work well with a
partition number offset (i.e., relative to the kernel partition), so
we'll be using a slightly different root UUID line.

NB: I've made this support specific to ip40xx for now, because I only
plan to support an IPQ4019-based AP that uses a Chromium-based
bootloader, but this image format can be used for essentially any
Chromebook, as well as the Google OnHub, a prior Chromium-based AP using
an IPQ8064 chipset.

[1]
ptgen: add Chromium OS kernel partition support
https://git.openwrt.org/?p=project/firmware-utils.git;a=commit;h=6c95945b5de973026dc6f52eb088d0943efa96bb

cros-vbutil: add Chrome OS vboot kernel-signing utility
https://git.openwrt.org/?p=project/firmware-utils.git;a=commit;h=8e7274e02fdc6f2cb61b415d6e5b2e1c7e977aa1

Signed-off-by: Brian Norris <computersforpeace@gmail.com>

computersforpeace committed:
ipq40xx: qcom_scm: Disable SDI at boot

See my upstream RFC of this:
https://lore.kernel.org/linux-arm-msm/20200721080054.2803881-1-computersforpeace@gmail.com/

This fixes warm boot (reboot) for Google WiFi devices using their
factory bootloader/firmware.

I may resend this upstream eventually.

Signed-off-by: Brian Norris <computersforpeace@gmail.com>

computersforpeace committed:
ipq40xx: qcom_scm: Fix cold boot address command

See my upstream questions:
https://lore.kernel.org/linux-arm-msm/20200913201608.GA3162100@bDebian/

This effectively reverts upstream Linux commit 13e77747800e ("firmware:
qcom: scm: Use atomic SCM for cold boot"), because Google WiFi boot
firmwares don't support the atomic variant.

This fixes SMP support for Google WiFi.

Signed-off-by: Brian Norris <computersforpeace@gmail.com>

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -77,7 +77,7 @@ endef
 define Build/cros-vboot
 	$(STAGING_DIR_HOST)/bin/cros-vbutil \
-		-k $@ -c "root=PARTUUID=%U/PARTNROFF=1" -o $@.new
+		-k $@ -c "root=PARTUUID=%U/PARTNROFF=1 rootwait" -o $@.new
```
This can go in the DTS using bootargs-append

computersforpeace replied:
Sure, done

robimarko commented:
Please drop the extra-features commit, it enables stuff that is already packaged as kernel modules for all of the boards.
Instead of doing that, just include the required kmods and package the Google FW stuff as kmods as well.

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+		ranges;
+
+		rsvd@87000000 {
+			/* Reserved for other subsystem */
```
What is this for?

computersforpeace replied:
Good question; I forgot I just borrowed that from the product DTS as-is.
Judging by some other context (the bootloader actually adds in some additional reserved-memory nodes of its own), I believe it's a conservative range carved out for some ATF/TrustZone usage. And in fact, Linux complains about some overlap at bootup, because the bootloader-provided regions sit right in the middle of this one.

I suspect it's OK to simply drop this node and defer to the bootloader instead. I plan to do that for the next revision.

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+
+/ {
+	model = "Google WiFi (Gale)";
+	compatible = "google,wifi", "google,gale-v2", "qcom,ipq4019";
```
Which one of these is needed?
The SoC one can be dropped for sure.

computersforpeace replied:
"google,wifi" is needed by OpenWRT for the naming scheme (matching this name with the board definitions in the Makefile), as discussed a ways back: http://patchwork.ozlabs.org/project/openwrt/patch/20200718205148.1743807-6-computersforpeace@gmail.com/.

"google,gale-v2" is needed by the bootloader, as it's looking for this product (code) name.

"qcom,ipq4019" is documented here: https://www.kernel.org/doc/Documentation/devicetree/bindings/arm/qcom.yaml

Why would I drop any of these?

robimarko replied:
You don't need the SoC compatible at all, take a look at other boards.

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+			status = "okay";
+		};
+
+		usb2: usb2@60f8800 {
```
Please use labels/phandles for all that have them and USB nodes have them

computersforpeace replied:
Ack, will do.

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+};
+
+&wifi0 {
+	status = "okay";
```
Are you sure that radios use generic BDF-s?

computersforpeace replied:
I'm not totally sure, but I'll go look at the original product SW to see.

BTW, in case you didn't notice, I documented the calibration data mechanism here: https://github.com/openwrt/openwrt/pull/9276/commits/1ba8cec6639ecc95e13b528e3e9d9776e122a241

I'm not sure how much the calibration data overlaps with (or overrides) BDF info.

chunkeey replied:
"per-calibration" is a loaded and still somewhat misleading term. It's not the same as the good-old' "calibration" from the ath5k/ath9k/qca9880 days. The firmware running in the QCA4019-IP-Block will pair the "pre-calibration" with the board data together to make the "calibration"... but it stays with the firmware. 

This is why vendors make the effort to put these board data files onto the devices. In most cases they are located in the rootfs in /lib/firmware/QCA4019 (or similar). Look there for the gold. The correct board data files can the big difference in how well a device's WIFI is performing.

robimarko replied:
Think of BDF-s as the generic WLAN configuration for that HW model while pre-cal is then made per each individual board.
And as @chunkeey mention, there is drastic performance and other differences if wrong BDF is used.

computersforpeace replied:
Honestly, I haven't really stress-tested the WiFi. It worked OK before, but I've now pulled the BDFs off the factory image and included those as a new 'variant=google,wifi'.

chunkeey commented on target/linux/ipq40xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -40,6 +40,19 @@ askey_do_upgrade() {
+	echo "Preparing to flash \"${board_dir}\" to /dev/mmcblk0p{1,2}"
+	ask_bool 0 "Abort" && exit 1
+
+	tar Oxf "${tar_file}" "${board_dir}/kernel" | dd of=/dev/mmcblk0p2 bs=1M
```
could this be a candidate for emmc_upgrade_tar()?

This is new, you can check it out @ https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/emmc.sh

computersforpeace replied:
Perhaps! Frankly I didn't actually spend much time on sysupgrade so it might not be working at all. I may give it another go and try out those helpers, so thanks for the pointer.

computersforpeace replied:
Awesome, this works.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+	pinctrl-names = "default";
+	status = "okay";
+
+	trusted-platform-module@20 {
```
you can call it "tpm". it's a valid "generic" [node name](https://devicetree-specification.readthedocs.io/en/v0.3/devicetree-basics.html#generic-names-recommendation)

computersforpeace replied:
Ack, will do.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+	lp55231@32 {
+		compatible = "national,lp5523";
+		reg = <0x32>;
+		clock-mode = [01];
```
The [device-tree binding example of the lp55xx](https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/leds-lp55xx.txt) prefers the `clock-mode = /bits/ 8 <1>;` notation. 

(was this copied from a de-compiled dtb? or is it the same in googles tree - not that this is important.)

computersforpeace replied:
The original DTS (not binary) used this notation. I'll change it though.

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+	cs-gpios = <&tlmm 45 0>;
+
+	spidev@0 {
+		compatible = "spidev";
```
Ok, this is special. I didn't know a "spidev" binding exists, so I looked and found this in [5.10](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/spi/spidev.c?h=v5.10#n740):


	/*
	 * spidev should never be referenced in DT without a specific
	 * compatible string, it is a Linux implementation thing
	 * rather than a description of the hardware.
	 */
	WARN(spi->dev.of_node &&
	     of_device_is_compatible(spi->dev.of_node, "spidev"),
	     "%pOF: buggy DT: spidev listed directly in DT\n", spi->dev.of_node);

(Newer kernels have this warning downgraded to a "dev_err").

Any ideas what device this is? Any unchecked features or chips left?

computersforpeace replied:
Good question, I forgot to revisit this. Poking around, I suspect it's the (unused, in production; and removed from some later HW spins) Zigbee radio. I believe that's supposed to be a SiliconLabs EM3581 chip? I don't see any device tree binding or driver for this kind of chip anywhere. Is it better to leave this as a "spidev" (perhaps with a comment)? Or remove it? Or partially include it in the DTS, but "disabled"?

chunkeey replied:
So, this "spidev" came in Google's own dts? In that case would I definitely document that strange oddity and not delete it.


As for  how exactly to document it... Good questions! The options you listed all sound fine. I probably would add mulltiline comment /* ... */ around the spi parent node (since from what I remember, that spidev is the only device there, right?) and add that "I suspect it's the (unused, in production; and removed from some later HW spins) Zigbee radio. I believe that's supposed to be a SiliconLabs EM3581 chip?"

computersforpeace replied:
Done

chunkeey commented on target/linux/ipq40xx/Makefile:
```diff
@@ -3,7 +3,7 @@ include $(TOPDIR)/rules.mk
 ARCH:=arm
 BOARD:=ipq40xx
 BOARDNAME:=Qualcomm Atheros IPQ40XX
-FEATURES:=squashfs fpu ramdisk nand
```
Hm, I've read the relevant commit message text... and I think you could make a case for a separate sub-target for that. I've done the same with the APM821xx / WD My Book Live. It lives there in a "sata" subtarget with boot-part and rootfs-part.

This would also allow you to "#ifdef" around that problematic SDI revert patch you have... You could make add a "Kconfig Symbol" that you only enable for the sub-target...

computersforpeace replied:
Thanks. This was actually a key question I was hoping to get feedback on, and I didn't get any with the mailing list posting.

I'll look at doing a subtarget. I agree that makes sense.

chunkeey replied:
hmm, I just remembered. There's also `of_machine_is_compatible("$machine_compatible")`. so, you could have the patches and skip the sub-target. hm

computersforpeace replied:
I've pushed a subtarget to address other concerns (additional USB drivers).

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+		reg = <0x32>;
+		clock-mode = [01];
+
+		chan0 {
```
I recently encountered this with another device (MPC85xx HiveAP330). It uses a LP5521 and has a multi-color led.

So upstream has this "multicolor"/"RGB" mode: 583ac0e11df7e29142837410a320db4654e4ff96 (see after the #else) lined up for RGB Leds. Unfortunately openwrt does not support this as of yet...

computersforpeace replied:
Can you provide a pointer to the "multicolor" documentation? I'm probably being dense, but I can't find any.

And: would you prefer I do the same approach (a preprocessed-away alternative)?

computersforpeace replied:
(FWIW, looking more I don't believe the multicolor stuff is really documented. But I figured out enough to get it working.)

Along the way, I noticed that between 5.4 (when I first got this all working) and 5.10, upstream changed the DT binding and broke my DTS. So I've pushed another version, fixing it up and including a multicolor (hidden by #if).

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+	cs-gpios = <&tlmm 12 0>;
```
You have included the gpio device-tree bindings header files. Can you please upgrade that `0` to `GPIO_ACTIVE_HIGH`.

computersforpeace replied:
Ack

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wifi.dts:
```diff
@@ -0,0 +1,402 @@
+	pinctrl-0 = <&spi_1_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+	cs-gpios = <&tlmm 45 0>;
```
This is another candidate for `GPIO_ACTIVE_HIGH`.

computersforpeace replied:
Ack

computersforpeace committed:
kernel: Package GOOGLE_FIRMWARE drivers

Useful for some Chromium OS based systems, like Google WiFi.

Signed-off-by: Brian Norris <computersforpeace@gmail.com>

computersforpeace commented:
> Please drop the extra-features commit, it enables stuff that is already packaged as kernel modules for all of the boards. Instead of doing that, just include the required kmods and package the Google FW stuff as kmods as well.

Commit 19fbca8769c581577d833eed5932b88e9849337c already notes why some if this must be done: "USB storage - this is the only way to install *-factory images without a lot of effort (serial console, etc.)". I can't make those kernel modules, because they are needed to load the filesystem from USB storage, and the filesystem is where we pull the kernel modules. But based on some other suggestions, I might make this a different subtarget. Would that alleviate some concerns about USB drivers?

For the GOOGLE_FIRMWARE drivers, I guess that's fine. I've written the kmod packaging now, and I can remove this from the main kernel in the next version.

For PSTORE: this isn't provided by any kmod-* package, and I'm not sure that's the best idea anyway. It's most useful for diagnosing kernel failures, and the earlier this is loaded the better. Plus, there are already 3 other targets that enable this.

> As far as SCM changes go, I am not for merging them as is, especially the change reverting atomic calls as it affects all of the boards.

Would you prefer a separate `compatible`, or the Kconfig/ifdef approach suggested by @chunkeey?

motolav commented:
> For PSTORE: this isn't provided by any kmod-* package, and I'm not sure that's the best idea anyway. It's most useful for diagnosing kernel failures, and the earlier this is loaded the better. Plus, there are already 3 other targets that enable this.

These PRs have the discussion on kmod over built-in for PSTORE  #5045  #5057

robimarko commented:
> > Please drop the extra-features commit, it enables stuff that is already packaged as kernel modules for all of the boards. Instead of doing that, just include the required kmods and package the Google FW stuff as kmods as well.
> 
> Commit [19fbca8](https://github.com/openwrt/openwrt/commit/19fbca8769c581577d833eed5932b88e9849337c) already notes why some if this must be done: "USB storage - this is the only way to install *-factory images without a lot of effort (serial console, etc.)". I can't make those kernel modules, because they are needed to load the filesystem from USB storage, and the filesystem is where we pull the kernel modules. But based on some other suggestions, I might make this a different subtarget. Would that alleviate some concerns about USB drivers?
> 
> For the GOOGLE_FIRMWARE drivers, I guess that's fine. I've written the kmod packaging now, and I can remove this from the main kernel in the next version.
> 
> For PSTORE: this isn't provided by any kmod-* package, and I'm not sure that's the best idea anyway. It's most useful for diagnosing kernel failures, and the earlier this is loaded the better. Plus, there are already 3 other targets that enable this.
> 
> > As far as SCM changes go, I am not for merging them as is, especially the change reverting atomic calls as it affects all of the boards.
> 
> Would you prefer a separate `compatible`, or the Kconfig/ifdef approach suggested by @chunkeey?

Then please make it a subtarget as there are board that have kernel size limitations in the generic target, this also allows using the SCM patches only on it.

computersforpeace committed:
firmware/ipq-wifi: Add Google Wifi board-2.bin

From a manufacturer's image (version R89-13729.57.27), with appopriate
',variant=' appended to the board names:

  $ .../qca-swiss-army-knife/tools/scripts/ath10k/ath10k-bdencoder \
        -i ./board-google_wifi.qca4019
  FileSize: 48596
  FileCRC32: 3966df5d
  FileMD5: d54161b0fb9e93691c4272649c37535a
  BoardNames[0]: 'bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=GO_GALE'
  BoardLength[0]: 12064
  BoardCRC32[0]: e117f336
  BoardMD5[0]: ea35e78c88a8571201da8b75edc9b881
  BoardNames[1]: 'bus=ahb,bmi-chip-id=0,bmi-board-id=21,variant=GO_GALE'
  BoardLength[1]: 12064
  BoardCRC32[1]: 6c751ec9
  BoardMD5[1]: 44cbc4ca6cb7141ba4249615f7065582
  BoardNames[2]: 'bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=GO_BREEZE'
  BoardLength[2]: 12064
  BoardCRC32[2]: 24fba117
  BoardMD5[2]: b4ac055b3ab67d5a6f5607a96af39a1f
  BoardNames[3]: 'bus=ahb,bmi-chip-id=0,bmi-board-id=21,variant=GO_BREEZE'
  BoardLength[3]: 12064
  BoardCRC32[3]: a3e16b2a
  BoardMD5[3]: 8b26cb285032314247304114b8ac50e7

Naming follows existing Google projects included in upstream board-2.bin
-- GO(ogle) prefix, an underscore (_), and the project code name, all in
caps.

Note that I only tested the "gale" model; the "breeze" model is a later
revision (same marketing name) with very small hardware changes but
otherwise using the same firmware image.

Submitted upstream here:

  ath10k-firmware: QCA4019: hw1.0: Add Google Wifi BDFs
  http://lists.infradead.org/pipermail/ath10k/2022-March/013465.html
  https://lore.kernel.org/ath10k/YjaNGW252Ls%2FyDw8@localhost/

Signed-off-by: Brian Norris <computersforpeace@gmail.com>

computersforpeace commented:
I've dropped the PSTORE stuff, since that can be a kernel module I guess, even if that's a bit suboptimal.

> Then please make it a subtarget as there are boards that have kernel size limitations in the generic target, this also allows using the SCM patches only on it.

I've pushed a subtarget, and I adjusted the patches to look for the machine compatible string. Hopefully this addresses all the concerns now.

computersforpeace commented:
Oops, I'm a bit new with Github PRs... there are "close" buttons for review comments, and there are "close" buttons that close the entire PR. I only meant to do the former!

Zebble commented:
Using @computersforpeace PR, I was able to build and successfully install openwrt on a Google Wifi device.  Thanks!  However, any changes made are lost after a reboot.  I'm a bit of an OpenWRT noob, so it's very possible I missed something in the build process but it did seem to work and created a squashfs image which should work?  Relevant errors in dmesg:

Sat Feb 26 03:18:03 2022 user.err kernel: [    8.251625] mount_root: failed to mount -t f2fs /dev/loop0 /tmp/overlay: No such device
Sat Feb 26 03:18:03 2022 user.info kernel: [    8.252102] mount_root: overlay filesystem has not been fully initialized yet
Sat Feb 26 03:18:03 2022 user.info kernel: [    8.259243] mount_root: switching to f2fs overlay
Sat Feb 26 03:18:03 2022 user.err kernel: [    8.265949] mount_root: switching to f2fs failed - fallback to ramoverlay

If this is the wrong place to post this, my apologies and would appreciate a point in the right direction.

Thanks again, great work!  So tired of the Google's Wifi devices forcing you to double-NAT if you want to use another router.  Looking forward to leveraging tools like DAWN with these pucks.

computersforpeace commented:
> Sat Feb 26 03:18:03 2022 user.err kernel: [ 8.251625] mount_root: failed to mount -t f2fs /dev/loop0 /tmp/overlay: No such device Sat Feb 26 03:18:03 2022 user.info kernel: [ 8.252102] mount_root: overlay filesystem has not been fully initialized yet Sat Feb 26 03:18:03 2022 user.info kernel: [ 8.259243] mount_root: switching to f2fs overlay Sat Feb 26 03:18:03 2022 user.err kernel: [ 8.265949] mount_root: switching to f2fs failed - fallback to ramoverlay

It's possible the default build isn't giving you kernel support for an appropriate on-disk filesystem for your overlay by default, and so it falls back to tmpfs (that loses your data upon reboot).

I'll poke around a bit, but I might need to add either `kmod-fs-f2fs` or `kmod-fs-ext4` to the `DEVICE_PACKAGES` listing. I'm a bit confused, because few other devices do this.

You could probably unblock your own builds by tweaking your `.config` (e.g., enable "Kernel module for F2FS filesystem support" in `make menuconfig`).

> If this is the wrong place to post this, my apologies and would appreciate a point in the right direction.

It's OK with me here! Glad to hear it's (mostly) working for you!

bobafetthotmail commented:
> I'll poke around a bit, but I might need to add either `kmod-fs-f2fs` or `kmod-fs-ext4` to the `DEVICE_PACKAGES` listing. I'm a bit confused, because few other devices do this.

kmods are kernel modules stored in rootfs partition. Rootfs partition cannot be read on boot without filesystem support in kernel so kmods can't help you.

You must enable filesystem support for your rootfs filesystem(s) of choice in the kernel config, see for example raspberry pi kernel config options where the ext4 and f2fs filesystems are enabled (as that is what is used in the two rootfs options) https://github.com/openwrt/openwrt/blob/ec6293febc244d187e71a6e54f44920be679cde4/target/linux/bcm27xx/bcm2708/config-5.4#L136

Zebble commented:
Good point @bobafetthotmail.  I have enabled ext4, f2fs and the loop block device for good measure as part of the kernel config and now changes persist a reboot!  I'm not sure if all 3 of those options were required but for now it works.  Now to keep playing and try DAWN.

computersforpeace commented:
Hi @bobafetthotmail,

> kmods are kernel modules stored in rootfs partition. Rootfs partition cannot be read on boot without filesystem support in kernel so kmods can't help you.

We're talking about overlay, not rootfs. (The default rootfs is squashfs.) You can most definitely use an overlay with a kernel module:

```
root@OpenWrt:~# lsmod | grep f2fs
f2fs                  360448  1 
root@OpenWrt:~# mount | grep f2fs
/dev/loop0 on /overlay type f2fs ([...snipped...])
```

But of course, building it into the kernel also works.

> You must enable filesystem support for your rootfs filesystem(s) of choice in the kernel config, see for example raspberry pi kernel config options where the ext4 and f2fs filesystems are enabled (as that is what is used in the two rootfs options)
> 
> https://github.com/openwrt/openwrt/blob/ec6293febc244d187e71a6e54f44920be679cde4/target/linux/bcm27xx/bcm2708/config-5.4#L136

Hmm,  I guess that works. Is ext4 overlay more common than f2fs? I see fewer that do that. But I also see targets that don't enable either. Frankly, I'm a bit new at finding "best practices" for OpenWrt.

chunkeey commented:
> Hmm, I guess that works. Is ext4 overlay more common than f2fs? I see fewer that do that. But I also see targets that don't enable either. Frankly, I'm a bit new at finding "best practices" for OpenWrt.

The best answer is probably: You need both, to have the (illusion of) choice for either.

As for good reasons? Well, I can think of the following:
 - f2fs is designed for eMMC/sdcards in mind.
 - ext4 comes as the stock filesystem on the devices from the vendor.
... Nuff' said.

Somewhat related but Off-and-On-topicy:

In the OpenWrt-World the main libs/executables are ideally always stored in the non-writeable squashfs rootfs. This is because OpenWrt's failsafe feature, the "restore to OpenWrt-Defaults by the press of the reset button" and the "firstboot" cli script are all designed around having the "changes" written to a separate overlaid fs that can simply be wiped which will return the device back to "OpenWrt-Defaults".

The way this "squashfs-root" works is that fstools's rootdisk sets up a loop-device in the "free space" after the squashfs until the end of the partition. It checks if this "free space" already contains a fs and if so, it mounts it as the overlay for / (this is also used by extroot). But if this "free space" is unpopulated or it finds the magic "firstboot/erase" marker it calls mkfs(.ext4/f2fs) to initialize it. Whenever it uses ext4 or f2fs depends on the size of the available remaining space. This is currently decided in a function called use_f2fs():

https://git.openwrt.org/?p=project/fstools.git;a=blob;f=libfstools/common.c;h=1b22d2cb19aefe1b928fad6045210b2656c1c16d;hb=922f1b3c091dbe12af4ba12a76ba8563418b688c#l105

`ret = size - offset > F2FS_MINSIZE;`

That `F2FS_MINSIZE` currently is `100ULL * 1024ULL * 1024ULL`. So for anything over 100MiB OpenWrt formats the overlay as f2fs. But if somehow some manages to fill up the squashfs rootfs to the point it that only 100MiB is left in the partition for the rootfs, then ext4 will be used.

computersforpeace commented:
@chunkeey Thanks for the info! That matches pretty much everything I discovered by reading through code. I'm planning to include `kmod-fs-f2fs` and `kmod-fs-ext4` in my next revision.

Semi-related note: I'm still having trouble with sysupgrade, where `rootfs_data` files don't always (ever?) get stashed/restored properly. It's possible they only accidentally get saved when the rootfs size doesn't change (and therefore the original rootfs_data just sits intact). This part is also tricky to debug, because even with serial console, the upgrade process doesn't print all that much by default.

chunkeey commented:
> Semi-related note: I'm still having trouble with sysupgrade, where rootfs_data files don't always (ever?) get stashed/restored properly. It's possible they only accidentally get saved when the rootfs size doesn't change (and therefore the original rootfs_data just sits intact). This part is also tricky to debug, because even with serial console, the upgrade process doesn't print all that much by default.

Did you have the problem with your previous, self-made sysupgrade? Or has this just pupping up with the emmc_* functions?

I don't have any devices that uses the EMMC, so this is very speculative. 

I too had issues like "sysupgrade not keeping the config" in the past with nand and nor. Most of these were related to block-size alignment. So I noticed earlier when looking at the that F2FS_MINSIZE, there's a ROOTDEV_OVERLAY_ALIGN (=64 KiB) right next to it which is used to setup the loop-device offset in:

https://lxr.openwrt.org/source/fstools/libfstools/rootdisk.c#L125 + L126

The new emmc_* functions in `package/base-files/files/lib/upgrade/emmc.sh` however use dd with "bs=512" to write into the EMMC_ROOT_DEV (which would apply to your case).

Could you try modifing emmc.sh's emmc_upgrade_tar() in either of the two ways?

1. aligning EMMC_ROOTFS_BLOCKS by 128 (64k / 512 = 128. So if the block count falls on the 128 boundry)

add `EMMC_ROOTFS_BLOCKS=$(((EMMC_ROOTFS_BLOCKS+127) & ~127))` before line 29 in emmc.sh

2. replace "bs=512" with "bs=64k" when relevant

This is in emmc_upgrade_tar in line 23, 29 and emmc_copy_config() in line 51. Just replace the "bs=512" with "bs=64k"

and check it again? (Be sure to include some verbose `echo`/`env/export` here and there to dump the EMMC_* variables that get set/calculated during sysupgrade... or make the bash extra verbose. `losetup -l dumps are also helpful)

(A common pit-fall: If you do the changes to your emmc.sh on your dev-machine to make a new image. Please don't forget to sync the new emmc.sh up with the file on the device before you test it. Otherwise the old emmc_upgrade_tar + emmc_copy_config will place the sysupgrade.tgz at the wrong location and you need to run the same new image twice through sysupgrade)

computersforpeace commented:
Hi @chunkeey,

Thanks a lot for the tips! They were exactly what I needed.

> Did you have the problem with your previous, self-made sysupgrade? Or has this just pupping up with the emmc_* functions?

I had similar problems with my self-made sysupgrade IIUC. But I believe you found the problem with both:

> Could you try modifing emmc.sh's emmc_upgrade_tar() in either of the two ways?
> 
> 1. aligning EMMC_ROOTFS_BLOCKS by 128 (64k / 512 = 128. So if the block count falls on the 128 boundry)

Yes indeed, the alignment seems to be the problem, and these suggestions fix things. I went with #1.

I've pushed a new series that should have f2fs and ext4 modules by default (to fix @bobafetthotmail's overlay fs report; thanks!), and with a fix for the emmc.sh helpers. I think it should all be ready to go.

chunkeey commented:
Thanks. I merged that one patch. As for the rest. Have you send the ipq40xx boardfiles to ath10k-devel yet?

computersforpeace commented:
> Thanks. I merged that one patch.

Thanks!

> As for the rest. Have you send the ipq40xx boardfiles to ath10k-devel yet?

No I haven't. I got the impression from somewhere on the openwrt lists that this isn't actually recommended. But I can't find such a reference at the moment, so if that's still the thing to do, I can give it a shot per:

https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/firmware/ipq-wifi/Makefile;h=9842725be29458c581b551f5f2e7460cc29b6818;hb=e8a0c55909427703b7c72ed46326531397287b20#l23
https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles

computersforpeace commented:
> > As for the rest. Have you send the ipq40xx boardfiles to ath10k-devel yet?

OK, I've rebased (to resolve the emmc.sh conflict), sent the BDFs [upstream](https://lore.kernel.org/ath10k/YjaNGW252Ls%2FyDw8@localhost/), and dropped the gen_vboot_image.sh script per the [discussion elsewhere](https://github.com/openwrt/openwrt/commit/e8a0c55909427703b7c72ed46326531397287b20#commitcomment-68923206).

chunkeey commented:
Merged. Thank you!

computersforpeace commented:
> Strange. All worked locally. Looks like the cros-vbutil is crashing/gets killed on the builders while packing the kernel.

Huh, I've never seen this. My impression with a "killed" status is that something else in the system is killing us, not that we crashed with, say, a memory or library error. (Those would be different exit statuses, right?) Is there any way to reproduce the builder setup?

This seems unlikely to matter, but technically there are some memory leaks in my tool, as I don't bother freeing up the intermediate buffers -- they're going to be freed shortly later, on exit, anyway. But if the buildbots are somehow putting excessive limits, I could, perhaps, reduce memory usage a bit by managing copies better. And I *am* loading entire kernel blobs into memory, which could probably be avoided...

Regarding SSL: no clue. I guess if I get info about the build environment, I could see if I can reproduce.

chunkeey commented:
> Huh, I've never seen this. My impression with a "killed" status is that something else in the system is killing us, not that we crashed with, say, a memory or library error. (Those would be different exit statuses, right?) Is there any way to reproduce the builder setup?

Yes, something like that. I'm not sure how, that's why I added both options. I know that a program can get "killed" by SIGILL when it is somehow executing a illegal instruction (seen this a lot with AVX512 on work - people want to run their HPC code on their shiny but shitty laptops for "debugging"). I'm not seeing that in your code. 

(It would be helpful to know if this "get killed" happen because of the input, or if any (reasonable) input causes it. I'll ask if something has some more insight. Again, it worked fine on my Ryzen laptop and Haswell machine)

> 
> This seems unlikely to matter, but technically there are some memory leaks in my tool, as I don't bother freeing up the intermediate buffers -- they're going to be freed shortly later, on exit, anyway. But if the buildbots are somehow putting excessive limits, I could, perhaps, reduce memory usage a bit by managing copies better. And I _am_ loading entire kernel blobs into memory, which could probably be avoided...
> 
> Regarding SSL: no clue. I guess if I get info about the build environment, I could see if I can reproduce.

From what I can tell, it's 3 MiB (for the zImage) - 10 (for the Initramfs) MiB on the HEAP. 
The calloc + malloc should handle this just fine.

chunkeey commented:
>  Is there any way to reproduce the builder setup?

Yes. The buildbots (workers) are all Debian 10 dockers. 

https://git.openwrt.org/?p=buildbot.git

Now, setting up the whole buildmaster+buildbot might be unnecessary?! At least I hope so.

From what I know, it should be possible to just run a buildworker environment:
(build-docker-images.sh uses:)
` docker build -t openwrtorg/buildworker -f docker/buildworker/Dockerfile . `

start up an interactive shell:
`docker run -it openwrtorg/buildworker /bin/bash`

In there download openwrt's git and start make menuconfig. Select the IPQ40xx, chromium sub-target (use multiple profiles, but select the "Google Wifi (Gale)" under Target devices. Enter  "Global build settings"  and go for the "Set build defaults for automatic builds (e.g. via buildbot)" this sets "CONFIG_BUILDBOT=y". Save and exit and run `make -j8`.

Let's see if this works or fails in the same way?

chunkeey commented:
It didn't fail either... But also:

https://downloads.openwrt.org/snapshots/targets/ipq40xx/chromium/

https://buildbot.openwrt.org/master/images/#/builders/79/builds/3

It worked? Builder was "sunshine-01" this time. Both failed builds came from "nomosphere-dock-03".
Both are on 3.4.0. This is getting weird.

(I'll re-enable chromium then. Will be out of town for a week though)

computersforpeace commented:
[ I meant to reply that I couldn't reproduce with the docker environment either. ]

> It worked? Builder was "sunshine-01" this time. Both failed builds came from "nomosphere-dock-03". Both are on 3.4.0. This is getting weird.

@ynezz helped me poke at this on the builder; we believe we found that it is indeed an openssl issue, where openssl is trying to use `getrandom(2)`, but fails because this particular builder has an ancient kernel without that syscall. I didn't get to the bottom of why openssl doesn't fall back to something like `/dev/random`, but anyway, I think the plan is to pull that machine off of building trunk.

> (I'll re-enable chromium then. Will be out of town for a week though)

Sounds good to me. Thanks for the help, both of you!

ynezz commented:
> (I'll re-enable chromium then. Will be out of town for a week though)

I hope, that you don't mind that I've re-enabled it just now. I just need to shuffle some buildworkers around and thus restart buildmaster and this would disable the builder for this chromium subtarget. Builders are not configured dynamically, they're configured when buildmaster instance (re)starts.
